components/camel-smb/src/main/java/org/apache/camel/component/smb/SmbProducer.java

```diff
@@ -0,0 +1,239 @@
+        LOGGER.debug("Normalize path {} to {}", file, result);
+        return result;
+    }
+
+    protected void connectIfNecessary(Exchange exchange) throws IOException {
+        Connection connection = smbClient.connect(getEndpoint().getHostname(), getEndpoint().getPort());
+
+        if (!loggedIn) {
+            LOGGER.debug("Not already connected/logged in. Connecting to: {}", getEndpoint());
+
+            AuthenticationContext ac = new AuthenticationContext(
+                    getEndpoint().getConfiguration().getUsername(),
+                    getEndpoint().getConfiguration().getPassword().toCharArray(),
+                    getEndpoint().getConfiguration().getDomain());
+            session = connection.authenticate(ac);
+
+            LOGGER.debug("Connected and logged in to: {}", getEndpoint());
+        }
+    }
+
+    public void disconnect() throws IOException {
+        loggedIn = false;
+        LOGGER.debug("Disconnecting from: {}", getEndpoint());
+
+        session.close();
+
+        session = null;
+    }
+
+    public void createDirectory(DiskShare share, java.io.File file) {
+        String parentDir = file.getParent();
+        SmbConfiguration configuration = getEndpoint().getConfiguration();
+
+        boolean dirExists = share.folderExists(parentDir);
+        if (!dirExists && configuration.isAutoCreate()) {
+            SmbFiles files = new SmbFiles();
+            files.mkdirs(share, normalize(parentDir));
+        }
+
+        if (!dirExists) {
+            throw new RuntimeCamelException("Directory " + parentDir + " does not exist on share " + share.toString());
+        }
+    }
+
+    private GenericFileExist determineFileExist(Exchange exchange) {
+        GenericFileExist gfe = exchange.getIn().getHeader(SmbConstants.SMB_FILE_EXISTS, GenericFileExist.class);
+        if (isEmpty(gfe)) {
+            gfe = getEndpoint().getConfiguration().getFileExist();
+        }
+
+        gfe = (gfe == null) ? GenericFileExist.Fail : gfe;
+
+        return gfe;
+    }
+
+    private void doFail(String path) {
+        throw new RuntimeCamelException("File " + path + " already exists, cannot create");
+    }
+
+    private void doIgnore(String path) {
+        // ignore but indicate that the file was written
+        LOGGER.debug("An existing file already exists: {}. Ignore and do not override it.", path);
+    }
+
+    private void removeFile(DiskShare share, java.io.File file) {
+        LOGGER.debug("An existing file already exists: {}. Removing it.", file.getName());
+        share.rm(file.getPath());
+    }
+
+    @Override
+    public void process(final Exchange exchange) {
+        String fileName = exchange.getIn().getHeader(Exchange.FILE_NAME, String.class);
+
+        SmbConfiguration configuration = getEndpoint().getConfiguration();
+        String path = (configuration.getPath() == null) ? "" : configuration.getPath();
+
+        LOGGER.debug("storeFile() fileName [" + fileName + "]");
```

Comment:
Use log markers here ... it's in the component's hot path.

Reply:
I ended up removing this statement - not sure it's necessary